src/Simple/JSON.purs

```diff
@@ -76,47 +79,58 @@ read = readImpl
 class ReadForeign a where
   readImpl :: Foreign -> F a
 
-instance readForeign :: ReadForeign Foreign where
+instance readForeignForeign :: ReadForeign Foreign where
   readImpl = pure
 
-instance readChar :: ReadForeign Char where
+instance readForeignChar :: ReadForeign Char where
   readImpl = readChar
 
-instance readNumber :: ReadForeign Number where
+instance readForeignNumber :: ReadForeign Number where
   readImpl = readNumber
 
-instance readInt :: ReadForeign Int where
+instance readForeignInt :: ReadForeign Int where
   readImpl = readInt
 
-instance readString :: ReadForeign String where
+instance readForeignString :: ReadForeign String where
   readImpl = readString
 
-instance readBoolean :: ReadForeign Boolean where
+instance readForeignBoolean :: ReadForeign Boolean where
   readImpl = readBoolean
 
-instance readArray :: ReadForeign a => ReadForeign (Array a) where
+instance readForeignArray :: ReadForeign a => ReadForeign (Array a) where
   readImpl = readElements <=< readArray
     where
       readElements xs = sequence $ readImpl <$> xs
 
-instance readNullOrUndefined :: ReadForeign a => ReadForeign (NullOrUndefined a) where
+instance readForeignNullOrUndefined :: ReadForeign a => ReadForeign (NullOrUndefined a) where
   readImpl = readNullOrUndefined readImpl
 
-instance readMaybe :: ReadForeign a => ReadForeign (Maybe a) where
+instance readForeignMaybe :: ReadForeign a => ReadForeign (Maybe a) where
   readImpl = map unNullOrUndefined <<< readImpl
 
-instance readNullable :: ReadForeign a => ReadForeign (Nullable a) where
+instance readForeignNullable :: ReadForeign a => ReadForeign (Nullable a) where
   readImpl o = withExcept (map reformat) $
     map toNullable <$> traverse readImpl =<< readNull o
     where
       reformat error = case error of
         TypeMismatch inner other -> TypeMismatch ("Nullable " <> inner) other
         _ -> error
 
-instance readStrMap :: ReadForeign a => ReadForeign (StrMap.StrMap a) where
+instance readForeignStrMap :: ReadForeign a => ReadForeign (StrMap.StrMap a) where
   readImpl = sequence <<< StrMap.mapWithKey (const readImpl) <=< readStrMap
 
-instance readRecord ::
+instance readForeignTuple :: (ReadForeign a, ReadForeign b) => ReadForeign (Tuple a b) where
+  readImpl = asTuple <=< readArray
+    where asTuple :: Array Foreign -> F (Tuple a b)
+          asTuple arr = case length arr of
```

Comment:
I think this code would be simpler by pattern matching on arr

```purescript
asTuple = case _ of
  [a, b] -> ...
  _ -> ...

Reply:
Of course, thanks.